component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,295 @@
+                    log.error("Retry attempt failed.", e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Will retrieve single Moesif key corresponding to given organization ID.
+     *
+     * @param orgID
+     * @return Moesif Key corresponding orgID
+     */
+    public String getMoesifKey(String orgID) {
+        String response;
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            response = callDetailResource(orgID);
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt of single moesif key fetch failed, retrying.", ex);
+
+            while (attempts > 0) {
+                attempts--;
+                try {
+                    Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    response = callDetailResource(orgID);
+                    return response;
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
+                }
+            }
+            response = null;
+        }
+        return response;
+    }
+
+    /**
+     * Will remove Moesif key corresponding to organization ID.
+     * Existing Moesif SDK client associated with the moesif key will be also removed.
+     *
+     * @param orgID
+     */
+    public void removeMoesifKeyFromMap(String orgID) {
+        String moesifKey = orgIDMoesifKeyMap.remove(orgID);
+        moesifKeyClientMap.remove(moesifKey);
+    }
+
+    /**
+     * Fetches all the available orgID - MoesifKey pairs from the microservice.
+     *
+     * @throws IOException
+     * @throws APICallException
+     */
+    private void callListResource() throws IOException, APICallException {
+        final URL obj;
+        try {
+            obj = new URL(MoesifMicroserviceConstants.LIST_URL);
+        } catch (MalformedURLException ex) {
+            log.error("Failed calling Moesif microservice. Attempted to call url: {}",
+                    MoesifMicroserviceConstants.LIST_URL, ex);
+            return;
+        }
+        String authHeaderValue = getAuthHeader(gaAuthUsername, gaAuthPwd);
+        HttpsURLConnection con = null;
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                try (BufferedReader in = new BufferedReader(
```

Comment:
@VirajSalaka : 

Let's do a try with resources here.
try(BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream(), "UTF-8"))) {
}

then you won't need in.close in the finally block

Comment : Ack